### ISSUE TITLE
How to define the class ODEFunc? What exactly does the forward function in the class ODEFunc mean?

### ISSUE THREAD

**Issue by RogerH2018** [StructuresComp/iMODE#1]

**Comment by liqiaofeng1990** [comment]
Hi, the forward function in the ODEFunc for the pendulum case is dong the following:
1. first construct an elastic potential energy field that is symmetric with respect to position. This is the "out" variable;
2. do backprop once to calculate an induced force field. This induced force field is automatically equivariant with respect to position. This is the "deriv" variable; 
3. construct the derivative vector for the equivalent first-order model of the pendulum. This is the "acc" variable.

Hope this helps.

**Comment by RogerH2018** [comment]
Hi, thanks for the response. I have one more question.
If the dynamics of the system are unknown and I only have trajectory information, how to define the class ODEFunc?
In other words, we didn't know the system was a pendulum. How to define the class ODEFunc?

**Comment by liqiaofeng1990** [comment]
Just to clarify that the above procedure can be used for any system that has a potential energy field, not only for a pendulum.

To be more general, if you don't know in advance that there is a potential energy field, you can do things pretty similarly: just take both the position and velocity (and the latent variable) as the input of your NN, and generate the acceleration as the output. Then you can concatenate it with the velocity, which will be the derivative vector for the equivalent first-order model.

To sum up, if you know a priori that there is a potential energy field, which generates the elastic force, you can use a NN to output the energy first and take backprop to generate the force/acceleration; if you don't know a priori that there is a potential energy field, you can use a NN to directly generate the force/acceleration.

**Comment by RogerH2018** [comment]
Thanks for the reply.
I understand that if I don't have any prior knowledge, the forward function in the class ODEFunc only needs to return the acceleration. This function does not require any physical information. Is this right?
In addition, if the trajectory is three-dimensional (x,y,z), how to calculate the acceleration? Could you give me an example in the forward function?
In my opinion, because there is no prior knowledge, I must construct a very complex neural network model.

**Comment by liqiaofeng1990** [comment]
Yes this function does not require physical information.

If the trajectory is 3D, the NN needs to output a 3D vector, i.e. the acceleration vector. For a quick example, if you assume 2 latent variables, the NN should take 8 inputs (x, y, z, v_x, v_y, v_z, two latent variables), and generates 3 outputs (acc_x, acc_y, acc_z).

We will keep updating the repository for other cases in the paper. Please stay on watch.
